memo: remove Memo() method from RelExpr interface

DrewKimball opened the pull request:
#### xform: check the memo before calling BuildLookupJoinLookupTableDistribution

This commit adds a check for nil and unoptimized memo to the implementation
of `GetLookupJoinLookupTableDistribution`. This is necessary because this
code path can be reached after the planner's optimizer has been reset during
formatting of a plan.

Informs #104009

Release note: None

#### opt: remove String() method from RelExpr

This patch removes the `String` method from the `RelExpr` interface.
This prepares to remove references to the memo from relational expressions.
For convenience, the memo now has the convenience methods `String()` and
`FormatExpr(expr)` for testing and debugging.

Informs #104009

Release note: None

#### opt: remove opttester references to RelExpr.Memo()

This patch removes references in the opttester framework to the `Memo`
method of `RelExpr`. This prepares for the removal of `Memo()`.

Informs #104009

Release note: None

#### opt/memo: remove references to RelExpr Memo() method

This patch removes various references to `RelExpr.Memo()` in the `memo`
package in preparation to remove the method entirely.

Informs #104009

Release note: None

#### opt: remove references to RelExpr Memo() method in optimizer and execbuilder

This patch removes various references to the `RelExpr.Memo()` method
throughout the optimizer and execbuilder logic.

Informs #104009

Release note: None

#### opt/props: remove references to RelExpr Memo() method

This commit removes references to the `RelExpr.Memo()` method throughout
the ordering and distribution props logic.

Informs #104009

Release note: None

#### memo: remove the RelExpr Memo() method

Since previous commits removed all references to the `RelExpr.Memo()`
method, it can now be removed entirely, and the reference to the memo
can be removed from `RelExpr` implementations.

Informs #104009

Release note: None

#### opt/memo: fix `(*statisticsBuilder).clear`

Release note: None

#### opt/memo: remove `(*Memo).EvalContext()`

Release note: None

#### opt/xform: clean up init

Release note: None

DrewKimball committed:
opt/props: remove references to RelExpr Memo() method

This commit removes references to the `RelExpr.Memo()` method throughout
the ordering and distribution props logic.

Informs #104009

Release note: None

DrewKimball committed:
opt: remove references to RelExpr Memo() method in optimizer and execbuilder

This patch removes various references to the `RelExpr.Memo()` method
throughout the optimizer and execbuilder logic.

Informs #104009

Release note: None

DrewKimball committed:
opt/memo: remove references to RelExpr Memo() method

This patch removes various references to `RelExpr.Memo()` in the `memo`
package in preparation to remove the method entirely.

Informs #104009

Release note: None

DrewKimball committed:
memo: remove the RelExpr Memo() method

Since previous commits removed all references to the `RelExpr.Memo()`
method, it can now be removed entirely, and the reference to the memo
can be removed from `RelExpr` implementations.

Informs #104009

Release note: None

DrewKimball committed:
opt: remove String() method from RelExpr

This patch removes the `String` method from the `RelExpr` interface.
This prepares to remove references to the memo from relational expressions.
For convenience, the memo now has the convenience methods `String()` and
`FormatExpr(expr)` for testing and debugging.

Informs #104009

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116243)
<!-- Reviewable:end -->

DrewKimball committed:
xform: check the memo before calling BuildLookupJoinLookupTableDistribution

This commit adds a check for nil and unoptimized memo to the implementation
of `GetLookupJoinLookupTableDistribution`. This is necessary because this
code path can be reached after the planner's optimizer has been reset during
formatting of a plan. In the case when the memo is nil or unoptimized, the
distribution will simply not be displayed.

Informs #104009

Release note: None

mgartner reviewed (commented):
Very nice! That was a lot of plumbing! :potable_water: 

> Reviewed 1 of 1 files at r1, 9 of 9 files at r2, 9 of 9 files at r3, 10 of 10 files at r4, 14 of 14 files at r5, 20 of 20 files at r6, 15 of 15 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-:-Nq8TYu14cq0gezwxP_4:bcpi3ma)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8LGXf9-1pKsc8tOJR):*
What is the impact of this? Will we show an incorrect (empty?) distribution in `EXPLAIN` output or elsewhere? Should we have some singleton that will print "unknown distribution" in this case?
___
*[`pkg/sql/opt/memo/expr_format.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8MQxP5UwtC3Mgrsr0:-Nq8MQxP5UwtC3Mgrsr1:b27x9zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/527160605ab1a81edfacc439af2730bfaa21bcfc/pkg/sql/opt/memo/expr_format.go#L161)):*
> ```Go
> 	// Memo must contain any expression that is formatted. It is not used for
> 	// scalar expressions. If it is nil for relational expressions, formatting may
> 	// result in a nil-pointer panic.
> ```

You could put a check for a nil memo in `formatRelational` that panics with a nicer message than "nil pointer exception". 
___
*[`pkg/sql/opt/memo/expr_name_gen.go` line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8Qc_qAfDHLdmBuu91:-Nq8Qc_qAfDHLdmBuu92:bz73s2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/18ad18f57d86bb2a844814c77aa89612157b1ac8/pkg/sql/opt/memo/expr_name_gen.go#L80)):*
> ```Go
> // expression. This function is used to generate consistent, unique names
> // for the columns in the table that will be created if the session
> // variable `save_tables_prefix` is non-empty.
> ```

TIL...
___
*[`pkg/sql/opt/memo/memo.go` line 567 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8MxuW9I8Zovfzn-Zg:-Nq8MxuW9I8Zovfzn-Zh:b-lbhdvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/527160605ab1a81edfacc439af2730bfaa21bcfc/pkg/sql/opt/memo/memo.go#L567)):*
> ```Go
> 
> // FormatExpr prints the given expression for testing and debugging.
> func (m *Memo) FormatExpr(expr opt.Expr) string {
> ```

It looks like this is only used in String above, so do we need it?
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 251 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8R1Ux4cgi1gekm1tB:-Nq8R1Ux4cgi1gekm1tC:b-mtzaf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/18ad18f57d86bb2a844814c77aa89612157b1ac8/pkg/sql/opt/memo/statistics_builder.go#L251)):*
> ```Go
> func (sb *statisticsBuilder) clear() {
> 	sb.evalCtx = nil
> 	sb.md = nil
> ```

I think we need to clear `sb.mem` here too: https://github.com/cockroachdb/cockroach/blob/7e3715c77b74398d581694fae5bddb93a65c9d26/pkg/sql/opt/memo/memo.go#L316-L319

Might be safest to change this to simply `*sb = statisticsBuilder{}`.
___
*[`pkg/sql/opt/norm/factory.go` line 300 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8SBFM3VqSqwypIXpt:-Nq8SBFM3VqSqwypIXpu:b-hmzf0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3baa7b9638607f1ec1d0e367739e420a76850f19/pkg/sql/opt/norm/factory.go#L300)):*
> ```Go
> // completely independent of it once CopyAndReplace has completed.
> func (f *Factory) CopyAndReplace(
> 	fromMemo *memo.Memo, from memo.RelExpr, fromProps *physical.Required, replace ReplaceFunc,
> ```

nit: update the comment and the sample usage section above
___
*[`pkg/sql/opt/testutils/opttester/opt_tester.go` line 640 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8O0q77WvsuxD6sKkm:-Nq8O0q77WvsuxD6sKkn:b8eead7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9907d1555632a1aabcf05762586d658d2ef6937/pkg/sql/opt/testutils/opttester/opt_tester.go#L640)):*
> ```Go
> 
> 	case "norm":
> 		mem, err := ot.OptNorm()
> ```

Would it be simpler in these cases to access the memo via `ot.f.Memo()` (or make an `ot.Memo()` method that does returns `ot.f.Memo()`)?
___
*[`pkg/sql/opt/xform/physical_props.go` line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8Ku75EBx2aUJVT72j:-Nq8Ku75EBx2aUJVT72k:b-px2e6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4a871541194a186689d1a52b73ccb8f579a32b7/pkg/sql/opt/xform/physical_props.go#L313)):*
> ```Go
> 		o, ok := optimizer.(*Optimizer)
> 		if !ok || o.mem == nil || !o.mem.IsOptimized() {
> 			return physicalDistribution
> ```

nit: here and in the other early returns we can `return physical.Distribution{}` to make it clear it's empty when you're reading this code


<!-- Sent from Reviewable.io -->

mgartner committed:
opt/memo: fix `(*statisticsBuilder).clear`

Release note: None

mgartner committed:
opt/memo: remove `(*Memo).EvalContext()`

Release note: None

mgartner committed:
opt/xform: clean up init

Release note: None

mgartner reviewed (commented):
I rebased this PR and addressed most of the comments I left. Some I ignored because I don't feel strongly about them anymore.

I also added 3 minor cleanup commits. Please take a look @DrewKimball!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-:-OJ-OO-A8hiRT3k3eoK8:b-88rndj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8LGXf9-1pKsc8tOJR:-OJ-ODYzD1_WPo2gLghM:bsprglz):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What is the impact of this? Will we show an incorrect (empty?) distribution in `EXPLAIN` output or elsewhere? Should we have some singleton that will print "unknown distribution" in this case?
</blockquote></details>

Ping on this question @DrewKimball, if it still makes sense.
___
*[`pkg/sql/opt/memo/expr_format.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8MQxP5UwtC3Mgrsr0:-OJ-LTgU-5KzSWtmZkTO:bfjoq95) ([raw file](https://github.com/cockroachdb/cockroach/blob/527160605ab1a81edfacc439af2730bfaa21bcfc/pkg/sql/opt/memo/expr_format.go#L161)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You could put a check for a nil memo in `formatRelational` that panics with a nicer message than "nil pointer exception".
</blockquote></details>

No need for this.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 251 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8R1Ux4cgi1gekm1tB:-OJ-MYkF5yB67qFnifYh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18ad18f57d86bb2a844814c77aa89612157b1ac8/pkg/sql/opt/memo/statistics_builder.go#L251)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think we need to clear `sb.mem` here too: https://github.com/cockroachdb/cockroach/blob/7e3715c77b74398d581694fae5bddb93a65c9d26/pkg/sql/opt/memo/memo.go#L316-L319

Might be safest to change this to simply `*sb = statisticsBuilder{}`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/norm/factory.go` line 300 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8SBFM3VqSqwypIXpt:-OJ-NLWDAhv9O-kIkpH2:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3baa7b9638607f1ec1d0e367739e420a76850f19/pkg/sql/opt/norm/factory.go#L300)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: update the comment and the sample usage section above
</blockquote></details>

done.
___
*[`pkg/sql/opt/xform/physical_props.go` line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8Ku75EBx2aUJVT72j:-OJ-O3ONDLbAUBgiRxDO:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4a871541194a186689d1a52b73ccb8f579a32b7/pkg/sql/opt/xform/physical_props.go#L313)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: here and in the other early returns we can `return physical.Distribution{}` to make it clear it's empty when you're reading this code
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
>I rebased this PR and addressed most of the comments I left. Some I ignored because I don't feel strongly about them anymore.

Thanks for taking the time to do that! I'll take a look today.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-:-OJ-QvAS1m381W-VyOm_:benuaid)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: remove opttester references to RelExpr.Memo()

This patch removes references in the opttester framework to the `Memo`
method of `RelExpr`. This prepares for the removal of `Memo()`.

Informs #104009

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8LGXf9-1pKsc8tOJR:-OJ-Rjm2Du1nUuXJi8fl:b-kxybhk):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Ping on this question @DrewKimball, if it still makes sense.
</blockquote></details>

We only display non-empty distribution, so it would mean we just don't display the distribution. I'll add that to the commit message.
___
*[`pkg/sql/opt/memo/memo.go` line 567 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8MxuW9I8Zovfzn-Zg:-OJ-USYe1TFnNsYgkuzl:bcsz54a) ([raw file](https://github.com/cockroachdb/cockroach/blob/527160605ab1a81edfacc439af2730bfaa21bcfc/pkg/sql/opt/memo/memo.go#L567)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It looks like this is only used in String above, so do we need it?
</blockquote></details>

It's nice to have for debugging, since you can do 
```
fmt.Println(m.FormatExpr(arbitraryExprInMemo))
```
___
*[`pkg/sql/opt/testutils/opttester/opt_tester.go` line 640 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116243#-Nq8O0q77WvsuxD6sKkm:-OJ0BntJ0wgDR6KuS4uA:b3k431u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9907d1555632a1aabcf05762586d658d2ef6937/pkg/sql/opt/testutils/opttester/opt_tester.go#L640)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Would it be simpler in these cases to access the memo via `ot.f.Memo()` (or make an `ot.Memo()` method that does returns `ot.f.Memo()`)?
</blockquote></details>

Nice idea. Referencing it via `ot.f.Memo()` simplifies the commit considerably.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTR and for the rebase!

bors r=mgartner
